schema/resource.go

```diff
@@ -13,64 +13,138 @@ type Resources []*Resource
+			r.data[i] = &cqtypes.Bool{
+				Status: cqtypes.Null,
+			}
+		case TypeInt:
+			r.data[i] = &cqtypes.Int8{
+				Status: cqtypes.Null,
+			}
+		case TypeFloat:
+			r.data[i] = &cqtypes.Float8{
+				Status: cqtypes.Null,
+			}
+		case TypeUUID:
+			r.data[i] = &cqtypes.UUID{
+				Status: cqtypes.Null,
+			}
+		case TypeString:
+			r.data[i] = &cqtypes.Text{
+				Status: cqtypes.Null,
+			}
+		case TypeByteArray:
+			r.data[i] = &cqtypes.Bytea{
+				Status: cqtypes.Null,
+			}
+		case TypeStringArray:
+			r.data[i] = &cqtypes.TextArray{
+				Status: cqtypes.Null,
+			}
+		case TypeIntArray:
+			r.data[i] = &cqtypes.Int8Array{
+				Status: cqtypes.Null,
+			}
+		case TypeTimestamp:
+			r.data[i] = &cqtypes.Timestamptz{
+				Status: cqtypes.Null,
+			}
+		case TypeJSON:
+			r.data[i] = &cqtypes.JSON{
+				Status: cqtypes.Null,
+			}
+		case TypeUUIDArray:
+			r.data[i] = &cqtypes.UUIDArray{
+				Status: cqtypes.Null,
+			}
+		case TypeInet:
+			r.data[i] = &cqtypes.Inet{
+				Status: cqtypes.Null,
+			}
+		case TypeInetArray:
+			r.data[i] = &cqtypes.InetArray{
+				Status: cqtypes.Null,
+			}
+		case TypeCIDR:
+			r.data[i] = &cqtypes.CIDR{
+				Status: cqtypes.Null,
+			}
+		case TypeCIDRArray:
+			r.data[i] = &cqtypes.CIDRArray{
+				Status: cqtypes.Null,
+			}
+		case TypeMacAddr:
+			r.data[i] = &cqtypes.Macaddr{
+				Status: cqtypes.Null,
+			}
+		case TypeMacAddrArray:
+			r.data[i] = &cqtypes.MacaddrArray{
+				Status: cqtypes.Null,
+			}
 		default:
-			sb.WriteString(fmt.Sprintf("%d", v))
+			panic(fmt.Errorf("unsupported type %s", r.Table.Columns[i].Type.String()))
 		}
 	}
-	return sb.String()
+	return &r
 }
 
-func (r *Resource) Get(key string) interface{} {
-	return r.Data[key]
+func (r *Resource) ToDestinationResource() DestinationResource {
+	dr := DestinationResource{
+		TableName: r.Table.Name,
+		Data:      r.data,
+	}
+	return dr
 }
 
-func (r *Resource) GetItem() interface{} {
-	return r.Item
+func (r *Resource) Get(columnName string) cqtypes.CQType {
+	index := r.Table.Columns.Index(columnName)
+	if index == -1 {
+		// we panic because we want to distinguish between code error and api error
+		// this also saves additional checks in our testing code
+		panic(columnName + " column not found")
+	}
+	return r.data[index]
 }
 
-func (r *Resource) Set(key string, value interface{}) error {
-	r.Data[key] = value
+// Set sets a column with value. This does validation and conversion to
+// one of concrete CQTypes. it returns an error just for backward compatibility
+// and panics in case it fails
+func (r *Resource) Set(columnName string, value interface{}) error {
+	index := r.Table.Columns.Index(columnName)
+	if index == -1 {
+		// we panic because we want to distinguish between code error and api error
+		// this also saves additional checks in our testing code
+		panic(columnName + " column not found")
+	}
+	if err := r.data[index].Set(value); err != nil {
+		panic(fmt.Errorf("failed to set column %s: %w", columnName, err))
```

Comment:
Should this be a panic? It seems like if we set a column and it fails, we should return an error but otherwise leave it empty.